embeddeddb/mysql/src/main/java/org/killbill/commons/embeddeddb/mysql/KillBillMariaDbDataSource.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2014-2018 Groupon, Inc
+ * Copyright 2014-2018 The Billing Project, LLC
+ *
+ * The Billing Project licenses this file to you under the Apache License, version 2.0
+ * (the "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at:
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.killbill.commons.embeddeddb.mysql;
+
+import java.sql.SQLException;
+import java.util.Properties;
+
+import org.mariadb.jdbc.MariaDbDataSource;
+import org.mariadb.jdbc.UrlParser;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Joiner;
+
+public class KillBillMariaDbDataSource extends MariaDbDataSource {
+
+    private static final Joiner.MapJoiner mapJoiner = Joiner.on('&').withKeyValueSeparator("=");
+
+    private String url;
+
+    private Boolean cachePrepStmts;
+    private Integer prepStmtCacheSize;
+    private Integer prepStmtCacheSqlLimit;
+    private Boolean useServerPrepStmts;
+
+    @Override
+    public void setUrl(final String url) throws SQLException {
+        this.url = url;
+        super.setUrl(url);
+    }
+
+    // See HikariDataSourceBuilder and https://github.com/brettwooldridge/HikariCP/wiki/MySQL-Configuration
+
+    public void setCachePrepStmts(final boolean cachePrepStmts) throws SQLException {
+        this.cachePrepStmts = cachePrepStmts;
+        updateUrlIfNeeded();
+    }
+
+    public void setPrepStmtCacheSize(final int prepStmtCacheSize) throws SQLException {
+        this.prepStmtCacheSize = prepStmtCacheSize;
+        updateUrlIfNeeded();
+    }
+
+    public void setPrepStmtCacheSqlLimit(final int prepStmtCacheSqlLimit) throws SQLException {
+        this.prepStmtCacheSqlLimit = prepStmtCacheSqlLimit;
+        updateUrlIfNeeded();
+    }
+
+    // Note! New default is now false (was true before 1.6.0)
+    public void setUseServerPrepStmts(final boolean useServerPrepStmts) throws SQLException {
+        this.useServerPrepStmts = useServerPrepStmts;
+        updateUrlIfNeeded();
+    }
+
+    private void updateUrlIfNeeded() throws SQLException {
+        if (url == null) {
+            final UrlParser urlParser = getUrlParser();
+            if (urlParser != null) {
+                url = urlParser.getInitialUrl();
+            }
+        }
+
+        if (url != null) {
+            setUrl(buildUpdatedUrl(url));
+        }
+    }
+
+    // No easy way to set MariaDB options at runtime besides updating the JDBC url
+    @VisibleForTesting
+    String buildUpdatedUrl(final String url) throws SQLException {
```

Comment:
Interesting. I am wondering if we should not even go one step further, as suggested by the HikariCP  link above:

```
dataSource.useLocalSessionState=true
dataSource.useLocalTransactionState=true
dataSource.rewriteBatchedStatements=true
dataSource.cacheResultSetMetadata=true
dataSource.cacheServerConfiguration=true
dataSource.elideSetAutoCommits=true
dataSource.maintainTimeStats=false
```

Also this post may be interested for this:  https://www.facebook.com/notes/mysql-at-facebook/jdbc-connection-overheads/10152027251270933/

Reply:
It looks like most of these only apply to the MySQL driver, not to the [MariaDB one](https://mariadb.com/kb/en/library/about-mariadb-connector-j/), except for `rewriteBatchedStatements` (which we don't want, as it overrides `useServerPrepStmts` to `false`).

That being said, there might be further work to be done to reduce roundtrips (`SET NAMES`, `SHOW VARIABLES`, etc.).